_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+
+Do dyspozycji mamy menu, z którego możemy przejść do:
+
+- `Public` - niezabezpieczonego route z `PublicComponent`
+- `Protected` - zabezpieczonego route z `ProtectedComponent`
+- `Keycloak Configuration` - niezabezpieczonego endpointu udostępniającego konfiguację Keycloak z backendu
+- `Backend Hello` - zabezpieczonego endpointu zwracającego tekst `Hello from the Backend!`
+- `Logout (backend)` - wylogowania z backendu
+- `Logout (frontend)` - przycisk widoczny tylko po zalogowaniu na frontendzie, który wylogowuje nas z frontendu
+
+Sprawdźmy więc czy mamy dostęp do `Keycloak Configuration` bez zalogowania:
+![Keycloak Configuration](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/keycloak_config.png)
+
+Wygląda na to, że nasz publicznie dostępny route i endpoint api działa poprawnie.
+
+Jeśli teraz przejdziemy do `Protected`, zostaniemy przekierowani na stronę logowania Keycloak:
+
+http://localhost:8180/auth/realms/**SpringBootAngular**/protocol/**openid-connect**/auth?client_id=**SpringBootAngularClient**&redirect_uri=**http%3A%2F%2Flocalhost%3A9082%2F%23%2F**&state=3e007783-4772-48dd-8b31-4bfe3cc9c42c&response_mode=fragment&response_type=code&scope=openid&nonce=023ab545-32dc-4bdc-9cbf-12cad1d0c944
+
+![Keycloak login](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/keycloak_login.png)
+
+Jak widzimy, adres zawiera informacje takie jak nazwa protokołu, realm, client czy adres na który mamy zostać przekierowani po zalogowaniu.
+
+Jeśli zalogujemy się użytkownikiem posiadającym rolę `user_role` (`user:password`), to zostaniemy przekierowani z powrotem do naszej aplikacji:
+
+![Public component](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/public_component_logged.png)
+
+Dodatkowo mamy jeszcze tylko przycisk `Logout (frontend)`. Jesteśmy teraz zalogowani, więc możemy przejść do `Protected`:
+
+![Protected component](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/protected_component.png)
+
+Zadziała również `Backend Hello`:
+
+![Backend Hello](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/backend_hello.png)
+
+Jeśli wylogujemy się teraz z frontendu - `Logout (frontend)`, to zniknie przycisk `Logout (frontend)` i nie będziemy mieli już dostępu do `Protected`.  
+Cały czas jednak będziemy mieli dostęp do `Backend Hello`. Dopiero kiedy wylogujemy się z backendu - `Logout (backend)`, to stracimy dostęp do `Backend Hello`.  
+Jednak jeśli będąc zalogowanym użytkownikiem, wylogujemy się z backendu a nie frontendu, to automatycznie zostaniemy również wylogowani z frontendu.
+
+Dlaczego?
+
+O tym jak działa Keycloak i tokeny którymi się posługujemy (a dokładniej protokół OpenID) nie jest tematem tego wpisu. Co do samego wylogowywania, to jest to temat na osobny wpis, ponieważ nie jest wcale taki prosty do zaimplementowania, szczególnie jeśli mówimy o rozproszonych systemach z wieloma instancjami aplikacji stojącymi za loadbalancerem.
```

Comment:
Czy to zdanie ma w związku sens w tym artykule?
"O tym jak działa Keycloak i tokeny którymi się posługujemy (a dokładniej protokół OpenID) nie jest tematem tego wpisu."

Natomiast odnośnie do drugiej części zdania to składam taką propozycję:
Odnośnie do kwestii wylogowywania, to temat nadaje się na osobny wpis, (ze względu na złożoność implementacji), szczególnie gdy mówimy o rozproszonych systemach z wieloma instancjami aplikacji stojącymi za loadbalancerem.

Reply:
Zdanie ma sens, ponieważ implementujemy tutaj coś, co wykorzystuje standard OpenID Connect. Chciałem zaznaczyć, że ma to związek z zasadą działania tego standardu.